lib/wallet/src/Cardano/Wallet/DB/Store/DeltaUTxO/Store.hs

```diff
@@ -0,0 +1,329 @@
+                , spentTxIns = case deltaUTxOValueSpent of
+                    Unspent -> spentTxIns
+                    Spent slot -> Map.insert txIn slot spentTxIns
+                , tip
+                , finality
+                , boot = boot <> case deltaUTxOValueCreation of
+                            Origin -> UTxO $ Map.singleton txIn txOut
+                            _at -> mempty
+                }
+      where
+        txIn = TxIn (getTxId deltaUTxOValueTxInTx) deltaUTxOValueTxInIx
+        txInSingleton = Set.singleton txIn
+        decodeTxOut = eitherToMaybe . deserializeTxOut . view lazy
+
+reverseMapOfSets :: Ord v => Map k (Set v) -> Map v k
+reverseMapOfSets m = Map.fromList $ do
+    (k, vs) <- Map.toList m
+    v <- Set.toList vs
+    pure (v, k)
+
+write :: WalletId -> UTxOHistory -> SqlPersistT IO ()
+write
+    wid
+    ( UTxOHistory
+            { history
+            , creationSlots
+            , spentSlots
+            , tip
+            , finality
+            , boot
+            }
+        ) = do
+        deleteWhere [DeltaUTxOSlotsWallet ==. wid]
+        insert_ $ DeltaUTxOSlots wid finality tip
+        deleteWhere [DeltaUTxOValueWalletId ==. wid]
+        insertMany_ $ do
+            (txIn@TxIn {inputId, inputIx}, txOut) <-
+                Map.assocs $
+                    unUTxO $
+                        history <> boot
+            creation <-
+                maybeToList $
+                    Map.lookup txIn $
+                        reverseMapOfSets creationSlots
+            let
+                spent = Map.lookup txIn $ reverseMapOfSets spentSlots
+            pure $
+                DeltaUTxOValue
+                    wid
+                    creation
+                    (maybe Unspent Spent spent)
+                    (TxId inputId)
+                    inputIx
+                    (encodeTxOut txOut)
+
+encodeTxOut = view strict . serializeTxOut
+encodeTxOut :: TxOut -> ByteString
+
+update :: WalletId -> Maybe UTxOHistory -> DeltaUTxOHistory -> SqlPersistT IO ()
+update wid = updateLoad (load wid) throwIO updateJust
+  where
+    updateJust :: UTxOHistory -> DeltaUTxOHistory -> SqlPersistT IO ()
+    updateJust old =
+        \case
+            AppendBlock newTip delta ->
+                constrainingAppendBlock (pure ()) old newTip $ do
+                    let
+                        reiceved = received delta
+                        spent = excluded delta
+                    insertMany_ $ do
+                        (TxIn {inputId, inputIx}, txOut) <-
+                            Map.assocs $
+                                unUTxO reiceved
+                        pure $
+                            DeltaUTxOValue
+                                wid
+                                (At newTip)
+                                Unspent
+                                (TxId inputId)
+                                inputIx
+                                (encodeTxOut txOut)
```

Comment:
I think that this is alright, but it does not replicate the exact same logic as in the Model:

```
Map.insert
    (At newTip)
    (received' `Set.difference` dom history)
```

In the model, the `dom history` is subtracted from the `received delta` before adding it, while this is not done in the database table. I think it's ok — as a precondition on `delta`, it is fair to assume that it doesn't create TxOuts that already exist. However, we need to be careful about this when generating `delta`.

Reply:
good point, should I remove the difference from the model ?

Reply:
I have kept the difference — but added a comment describing it.